src/imitation/util/serialize.py

```diff
@@ -20,10 +20,20 @@ class Serializable(ABC):
   @abstractmethod
   def load(cls, directory):
     """Load object plus weights from directory."""
+    with open(os.path.join(directory, 'loader'), 'rb') as f:
+      load_cls = pickle.load(f)
+    return load_cls.load(directory)
```

Comment:
The `Serializable.load` path ends up calling the actual implementation of `DiscrimNet.load`. Maybe there's something I don't understand about how name resolution works for `DiscrimNet.load`, but I thought that `DiscrimNet.load` should lead to the `def load(cls, ...)` defined in `DiscrimNet`.

Reply:
This is working as intended, but on reflection I think the design may be too confusing.

The desiderata here are:
  - I should be able to load any `Serializable` object via a single method without worrying about what type it is.
  - Different type objects should be able to support arbitrary serialization/deserialization logic, due to the heterogenity of things we might want to save. (For example, I have some reward models that wrap other reward models -- so serializing them means first serializing the inner model, then the outer model.)

I implemented this by:
  - Making `Serializable.save` save the type of the object into `loader`. Concrete implementations of this call it via super().
  - Concrete implementations of `Serializable` *must* override `load`, since it's abstract. (Otherwise `Serializable.load` could end up calling itself, which would be bad.)
  - `Serializable.load` read from `loader` and calls the type-specific `load`. This works since any concrete type must have its own `load`.

The confusion here is that `DiscrimNet` is itself an abstract class, and there's no `DiscrimNet.load`, since `DiscrimNetAIRL` and `DiscrimNetGAIL` do quite different things. So in this case `DiscrimNet.load` just goes through to `Serializable.load`, which then calls `DiscrimNetAIRL.load` or `DiscrimNetGAIL.load` as appropriate.

I'd appreciate alternative design suggestions: calling an abstract method, while legal Python, does feel odd. One alternative would be to just add a method `load_serialize` which does what `Serializable.load` currently does. I do like that e.g. `DiscrimNet.load` self-documents that we expect it to be a `DiscrimNet` (and potentially enforces this via type-checking), which would be lost with a generic helper method.

Reply:
Note this is also true for `reward_net.RewardNet.load`, as there is no `load` method implemented in `RewardNet` (also abstract).

Reply:
Thanks for writing up such a thorough explanation!

Yesterday when I was reading through the code I mistakenly confused the `DiscrimNetAIRL.load` for a `DiscrimNet.load` implementation. So I thought that `DiscrimNet.load` had an implementation, which was somehow being skipped in favor of `Serializable.load`. (Good to know now that my understanding of Python name resolution isn't broken).